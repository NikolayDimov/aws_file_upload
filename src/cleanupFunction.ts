import { DynamoDBClient, ScanCommand, DeleteItemCommand } from "@aws-sdk/client-dynamodb";
import * as AWS from "aws-sdk";

const s3 = new AWS.S3();
const dynamoClient = new DynamoDBClient({});

const TABLE_NAME = process.env.TABLE_NAME!;
const BUCKET_NAME = process.env.BUCKET_NAME!;

export const cleanupFunction = async (event: any) => {
    const now = Math.floor(Date.now() / 1000);

    // Query DynamoDB for expired files
    const expiredFiles = await dynamoClient.send(
        new ScanCommand({
            TableName: TABLE_NAME,
            FilterExpression: "expirationTime <= :now",
            ExpressionAttributeValues: { ":now": { N: now.toString() } },
        })
    );

    for (const file of expiredFiles.Items || []) {
        const fileKey = file.fileKey.S;

        if (fileKey) {
            // Delete file from S3
            await s3
                .deleteObject({
                    Bucket: BUCKET_NAME,
                    Key: fileKey,
                })
                .promise();

            console.log(`Deleted file: ${fileKey}`);

            // Delete metadata from DynamoDB
            await dynamoClient.send(
                new DeleteItemCommand({
                    TableName: TABLE_NAME,
                    Key: { fileKey: { S: fileKey } },
                })
            );

            console.log(`Deleted metadata for file: ${fileKey}`);
        }
    }

    return {
        statusCode: 200,
        body: "Cleanup function processed successfully",
    };
};
