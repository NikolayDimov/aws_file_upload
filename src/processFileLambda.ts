import { DynamoDBClient, PutItemCommand } from "@aws-sdk/client-dynamodb"; // Importing low-level DynamoDB client and command
import * as AWS from "aws-sdk";

const s3 = new AWS.S3();
const dynamoClient = new DynamoDBClient({});

const TABLE_NAME = process.env.TABLE_NAME;

export const processFileLambda = async (event: any) => {
    for (const record of event.Records) {
        const bucket = record.s3.bucket.name;
        const key = record.s3.object.key;
        const size = record.s3.object.size;

        const expirationTime = Math.floor(Date.now() / 1000) + 30 * 60;
        const fileExtension = key.split(".").pop();

        // Construct the PutItem parameters for low-level API
        const putItemParams = {
            TableName: TABLE_NAME,
            Item: {
                fileKey: { S: key },
                expirationTime: { N: expirationTime.toString() },
                fileExtension: { S: fileExtension || "" },
                fileSize: { N: size.toString() },
                uploadTime: { S: new Date().toISOString() },
            },
        };

        console.log("DynamoDB Params:", putItemParams);

        // Use PutItemCommand to insert the metadata into DynamoDB
        try {
            await dynamoClient.send(new PutItemCommand(putItemParams));
            console.log(`Metadata for ${key} stored successfully`);
        } catch (error) {
            console.error("Error storing metadata for", key, error);
        }
    }
};
