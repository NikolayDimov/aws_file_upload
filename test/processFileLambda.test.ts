import * as cdk from "aws-cdk-lib";
import { Stack } from "aws-cdk-lib";
import { Table, AttributeType } from "aws-cdk-lib/aws-dynamodb";
import { Topic } from "aws-cdk-lib/aws-sns";
import { processFileLambda } from "../src/processFileLambda";
import "jest-cdk-snapshot";
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { SNSClient } from "@aws-sdk/client-sns";

// Mock AWS SDK clients
jest.mock("@aws-sdk/client-dynamodb", () => ({
    DynamoDBClient: jest.fn().mockImplementation(() => ({
        send: jest.fn(),
    })),
    PutItemCommand: jest.fn(),
}));

jest.mock("@aws-sdk/client-sns", () => ({
    SNSClient: jest.fn().mockImplementation(() => ({
        send: jest.fn(),
    })),
    PublishCommand: jest.fn(),
}));

describe("handler function tests with CDK snapshot", () => {
    let dynamoClient: DynamoDBClient;
    let snsClient: SNSClient;

    beforeEach(() => {
        process.env.TABLE_NAME = "ErrorTable";
        process.env.TOPIC_ARN = "arn:aws:sns:us-east-1:123456789012:ErrorTopic";

        dynamoClient = new DynamoDBClient({});
        snsClient = new SNSClient({});

        (dynamoClient.send as jest.Mock).mockImplementation(() => Promise.resolve({}));
        (snsClient.send as jest.Mock).mockImplementation(() => Promise.resolve({}));
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test("should call DynamoDB for invalid JSON", async () => {
        const event = {
            Records: [
                {
                    s3: {
                        bucket: { name: "my-bucket" },
                        object: { key: "invalid-file.jpg", size: 5678 },
                    },
                },
            ],
        };

        await processFileLambda(event);

        const stack = new Stack();
        new Table(stack, "ErrorTable", {
            partitionKey: { name: "id", type: AttributeType.STRING },
        });

        new Topic(stack, "ErrorTopic");

        expect(stack).toMatchCdkSnapshot();
    });

    test("should send an SNS notification for valid JSON", async () => {
        const event = {
            Records: [
                {
                    s3: {
                        bucket: { name: "my-bucket" },
                        object: { key: "valid-file.jpg", size: 5678 },
                    },
                },
            ],
        };

        await processFileLambda(event);

        const stack = new Stack();
        new Table(stack, "ErrorTable", {
            partitionKey: { name: "id", type: AttributeType.STRING },
        });

        new Topic(stack, "ErrorTopic");

        expect(stack).toMatchCdkSnapshot();
    });
});
