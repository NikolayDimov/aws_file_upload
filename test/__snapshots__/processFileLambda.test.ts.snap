// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler function tests with CDK snapshot should call DynamoDB for invalid JSON 1`] = `
{
  "Resources": {
    "ErrorTableA27F7F12": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "ErrorTopicA0904A23": {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;

exports[`handler function tests with CDK snapshot should send an SNS notification for valid JSON 1`] = `
{
  "Resources": {
    "ErrorTableA27F7F12": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "ErrorTopicA0904A23": {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
